Sources/Sentry/SentryFramesTracker.m

```diff
@@ -106,26 +125,45 @@ - (void)displayLinkCallback
     CFTimeInterval slowFrameThreshold = 1 / (actualFramesPerSecond - 1);
 
     CFTimeInterval frameDuration = lastFrameTimestamp - self.previousFrameTimestamp;
-    self.previousFrameTimestamp = lastFrameTimestamp;
 
     if (frameDuration > slowFrameThreshold && frameDuration <= SentryFrozenFrameThreshold) {
         atomic_fetch_add_explicit(&_slowFrames, 1, SentryFramesMemoryOrder);
-    }
-
-    if (frameDuration > SentryFrozenFrameThreshold) {
+#    if SENTRY_TARGET_PROFILING_SUPPORTED
+        [self recordTimestampStart:@(self.previousFrameTimestamp) end:@(lastFrameTimestamp)];
+#    endif // SENTRY_TARGET_PROFILING_SUPPORTED
+    } else if (frameDuration > SentryFrozenFrameThreshold) {
         atomic_fetch_add_explicit(&_frozenFrames, 1, SentryFramesMemoryOrder);
+#    if SENTRY_TARGET_PROFILING_SUPPORTED
+        [self recordTimestampStart:@(self.previousFrameTimestamp) end:@(lastFrameTimestamp)];
+#    endif // SENTRY_TARGET_PROFILING_SUPPORTED
     }
 
     atomic_fetch_add_explicit(&_totalFrames, 1, SentryFramesMemoryOrder);
+
+    self.previousFrameTimestamp = lastFrameTimestamp;
+}
+
+#    if SENTRY_TARGET_PROFILING_SUPPORTED
+- (void)recordTimestampStart:(NSNumber *)start end:(NSNumber *)end
+{
+    if (self.currentTracer.isProfiling) {
+        [self.frameTimestamps addObject:@{ @"start_timestamp" : start, @"end_timestamp" : end }];
+    }
 }
+#    endif // SENTRY_TARGET_PROFILING_SUPPORTED
```

Comment:
`l`: If we do it like this, we could remove the target conditionals above and make this an empty function when profiling is not supported, which the compiler should ditch when optimizing the code.
```suggestion
- (void)recordTimestampStart:(NSNumber *)start end:(NSNumber *)end
{
#    if SENTRY_TARGET_PROFILING_SUPPORTED
    if (self.currentTracer.isProfiling) {
        [self.frameTimestamps addObject:@{ @"start_timestamp" : start, @"end_timestamp" : end }];
    }
#    endif // SENTRY_TARGET_PROFILING_SUPPORTED
}

```

Reply:
Whaat? Stupid ObjC compiler. TIL.